### ISSUE TITLE
How to copy code from external buffer to CodeHolder?

### ISSUE THREAD

**Issue by sakjain92** [asmjit/asmjit#304]
My aim is to cache the code inside CodeHolder (dumped via x86Assembler) and then reuse it in the future multiple times? 

i.e.
Writing to cache: CodeHolder -> buffer (via `copyFlattenedData()`) -> file on disk
Reading from cache: file on disk -> buffer -> CodeHolder (via ? function) -> mmap to executable region (via `relocateToBase()`) 

My question is:
1) Is this possible?
2) If so, how do I dump data in a buffer to a CodeHolder, in such a way that I can then call relocateToBase() on this new CodeHolder? Can you give me an example?

**Comment by kobalicek** [comment]
To do this successfully, you would have to:

  - Serialize everything `CodeHolder` contains, which means storing sections, their buffers, and other data like labels, label links, and relocation information.
  - Never use absolute addresses to functions that live outside of the generated code, like memcpy, for example. You would have to  make a function table or something else in that sense so another process can link it properly.

Basically, you would have to do in some simplified way what linkers do. There is no support out of box for that feature and I'm not able to provide you an example. If you really need this feature you would have to implement it, and possibly create issues regarding missing pieces in AsmJit to make that possible.

Hope that helps, let me know if you have further questions.

**Comment by sakjain92** [comment]
Thanks for the quick reply.

I was thinking about this for now.
Writing to cache:
1) Emit stuff into CodeHolder
1a) For external functions which are not part of generated code, I keep the address of these functions in a separate section, say `.externalFunctions` and to call these functions, I use `call [address in .externalFunctions section]` in the generated code (Sort of similar to how PIC program works I think)
2) Resolve all relocation information using `relocateToBase(<random address>)`. (I would think that there would be no need to relocate since every address should be relative (except for the hard-coded external function addresses). Calling `relocateToBase()` I am hoping will resolve all internal labels atleast)
3) Copy everything out of CodeHolder using  `copyFlattenedData()`
4) Dump this data into a file. Add some metadata in the file such as offset and size of `.externalFunctions` and what external function it contains and at what offset inside this section.

Reading from cache:
1) Read data from file
2) Seek to offset of `.externalFunctions` section. Update the function addresses (Address Space Layout Randomization will cause function addresses to be changed) to the current addresses of the external functions.
3) Copy data into a Read-only+Executable section via mmap
4) Jump to start of this region.

These steps are similar to what you mentioned in your answer. But are these steps enough or am I missing something?

**Comment by kobalicek** [comment]
Yeah that could work I think

**Comment by connorjclark** [comment]
@sakjain92 if you happened to ever implement your approach, is it something you can share?
